### ISSUE TITLE
gRPC AttributeExtractors should have access to gRPC Metadata passed to TracingClientCall.start()

### ISSUE THREAD

**Issue by evantorrie** [open-telemetry/opentelemetry-java-instrumentation#6123]
**Is your feature request related to a problem? Please describe.**

TL;DR; Custom AttributeExtractors added to instrumented gRPC client calls lack access to gRPC metadata sent in call.

Currently, the gRPC instrumentation constructs a `GrpcRequest` object at the beginning of `interceptCall()` at which point the gRPC Metadata is unset - [set to `null` in the code](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/efd40f8cdf73f33b8a055359d7fe8bed5afaf54c/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6/TracingClientInterceptor.java#L46).

gRPC Metadata for outgoing client calls *is* available at the point of `TracingClientCall.start(Listener<RESPONSE> responseListener, Metadata header)`. However, the gRPC request object which is passed to `instrumenter.end()` never has its corresponding Metadata private field updated.

**Describe the solution you'd like**

`instrumenter.end()` calls each AttributeExtractor's `onEnd()` method. The `request` object passed to these methods should include an updated version of the metadata sourced from the metadata parameter passed to `TracingClientCall.start()`.

**Describe alternatives you've considered**

Duplicating most of the code from the grpc Instrumentation simply to ensure that the `GrpcRequest` object is updated with the appropriate metadata to make it available to our own AttributeExtractors.

**Additional context**

This issue was raised at the [May 10th Java SIG](https://docs.google.com/document/d/1WK9h4p55p8ZjPkxO75-ApI9m0wfea6ENZmMoFRvXSCw/edit#heading=h.wtapn135xim7).

**Comment by trask** [comment]
hi @evantorrie! it would be great if you are able to send a PR for this

**Comment by laurit** [comment]
Resolved with https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6125
